File: BinaryIndexedTree/FenwickTree/FenwickTree.java
class FenwickTree{
    private int[] array;
    public FenwickTree(int n){
        array=new int[n];
    }
    
    public void update(int i, int val){
        while (i<array.length){
            array[i]+=val;
            i += lowbit(i);
        }
    }

    public int query(int i){
        int sum=0;
        if (i>=array.length)
            return sum;
        while (i>0){
            sum += array[i];
            i -= lowbit(i);
        }
        return sum;
    }

    private int lowbit(int x){
        return x&(-x);
    }
}

File: BinaryIndexedTree/FenwickTree/FenwickTree.py
class FenwickTree:
    # Use for partial sum of an array, and the array is MODIFIABLE.
    def __init__(self, N: int):
        self.tree = [0]*(N+1)

    def __lowbit(self, x: int):
        return x&(-x)

    def update(self, i: int, delta: int):
        size = len(self.tree)
        while i<size:
            self.tree[i]+=delta
            i+=self.__lowbit(i)
    
    def query(self, i: int):
        count=0
        while i>0:
            count += self.tree[i]
            i-=self.__lowbit(i)
        return count
